core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java

```diff
@@ -124,16 +125,54 @@ public String generateLoadBalancerEndpoint(Stack stack) {
         }
     }
 
+    /*
+     *  follows these peferences:
+     *   public fqdn
+     *   public DNS
+     *   public IP
+     *   private FQDN
+     *   private DNS
+     *   private IP
+     */
     public String getLoadBalancerUserFacingFQDN(Long stackId) {
         Set<LoadBalancer> loadBalancers = loadBalancerPersistenceService.findByStackId(stackId).stream()
-            .filter(lb -> StringUtils.isNotEmpty(lb.getFqdn()))
-            .collect(Collectors.toSet());
-        LoadBalancer loadBalancer = loadBalancers.stream()
-            .filter(lb -> LoadBalancerType.PUBLIC.equals(lb.getType()))
-            .findAny()
-            .orElseGet(() -> loadBalancers.stream().findAny().orElse(null));
+                .filter(lb -> StringUtils.isNotEmpty(lb.getDns()) ||
+                        StringUtils.isNotEmpty(lb.getIp()) || StringUtils.isNotEmpty(lb.getFqdn()))
+                .collect(Collectors.toSet());
+        if (loadBalancers.isEmpty()) {
+            return null;
+        }
+        Iterator<LoadBalancer> iterator = loadBalancers.iterator();
+        LoadBalancer best = iterator.next();
+        while (iterator.hasNext()) {
+            LoadBalancer balancer = iterator.next();
+            if (!best.getType().equals(balancer.getType())) {
+                if (best.getType().equals(LoadBalancerType.PRIVATE)) {
+                    best = balancer;
+                }
+                continue;
+            }
+            if (StringUtils.isNotEmpty(balancer.getFqdn()) && StringUtils.isEmpty(best.getFqdn())) {
```

Comment:
This if statement will only be hit if we have two load balancers in the set that are the same type. Functionally, that shouldn't happen. The load balancers in the set will always have different types, because we only create new load balancer objects if there isn't one of the desired type already saved. (This class also has does the initial load balancer creation, and that check is done in the `createLoadBalancerIfNotExists` method.) I understand what you were doing with this clause, but it's not a valid case due to how the code is written.

Reply:
This was a piece of information I was missing, Is it correct to say there will only ever be at most 1 Public and 1 Private load balancer returned for a given stackId?

Reply:
Correct.